File: PathValueResolvers/PathValueResolverRegistry.cpp
// Copyright (c) 2010-2017 Fabric Software Inc. All rights reserved.

#include "PathValueResolverRegistry.h"
 
using namespace FabricUI;
using namespace FabricCore;
using namespace PathValueResolvers;

bool PathValueResolverRegistry::s_instanceFlag = false;
PathValueResolverRegistry* PathValueResolverRegistry::s_registry = 0;

PathValueResolverRegistry::PathValueResolverRegistry()
{
}

PathValueResolverRegistry::~PathValueResolverRegistry()
{
  s_instanceFlag = false;
}

PathValueResolverRegistry* PathValueResolverRegistry::GetRegistry()
{
  if(s_instanceFlag == false)
  {
  	s_registry = new PathValueResolverRegistry();
  	s_instanceFlag = true;
  }
  return s_registry;
}

void PathValueResolverRegistry::registerResolver(
	const QString name,
  PathValueResolver *resolver)
{
	if(!hasResolver(name))
		m_registeredSolvers[name] = resolver;
}

void PathValueResolverRegistry::unregisterResolver(
  const QString name)
{
	if(hasResolver(name))
		m_registeredSolvers.remove(name);
}

bool PathValueResolverRegistry::hasResolver(
  const QString name)
{
	return m_registeredSolvers.count(name);
}

PathValueResolver* PathValueResolverRegistry::getResolver(
  const QString name)
{
	return hasResolver(name) 
		? m_registeredSolvers[name]
		: 0;
}
 
bool PathValueResolverRegistry::knownPath(
	RTVal pathValue)
{
	foreach(PathValueResolver* resolver, m_registeredSolvers)
	{
		if(resolver->knownPath(pathValue))
			return true;
	}
	return false;
}

QString PathValueResolverRegistry::getType(
	RTVal pathValue)
{
	foreach(PathValueResolver* resolver, m_registeredSolvers)
	{
		if(resolver->knownPath(pathValue))
			return resolver->getType(pathValue);
	}
	return "";
}

void PathValueResolverRegistry::getValue(
	RTVal pathValue)
{
	foreach(PathValueResolver* resolver, m_registeredSolvers)
	{
		if(resolver->knownPath(pathValue))
			resolver->getValue(pathValue);
	}
}

void PathValueResolverRegistry::setValue(
	RTVal pathValue)
{
	foreach(PathValueResolver* resolver, m_registeredSolvers)
	{
		if(resolver->knownPath(pathValue))
			resolver->setValue(pathValue);
	}
}

File: Python/Canvas/Commands/PathValueResolverRegistry.py
#
# Copyright (c) 2010-2017 Fabric Software Inc. All rights reserved.
#

from FabricEngine.FabricUI import Commands as CppCommands

class PathValueResolverRegistry(CppCommands.PathValueResolverRegistry):

    """ 
    """
    def __init__(self): 
        """ Initializes the PathValueResolverRegistry.
        """
        super(PathValueResolverRegistry, self).__init__()
        # There is no "new" in python, we need to own the resolver 
        # created in Python. They are referenced in the C++ registry. 
        self.__flatResolverStack = {}

    def registerResolver(self, name, resolverType, userData = None):
        """ Registers a Python PathValue resolver 'resolverType' under the name
            "name" and creates it. Raises an exception if the name is empty or 
            the resolverType is not a PathValueResolver.
        """
        if not name:
          raise Exception('PathValueResolverRegistry.registerResolver, error: name is empty') 

        if not issubclass(resolverType, CppCommands.BaseCommand):
            raise Exception(
                "PathValueResolverRegistry.registerResolver, error: '" + 
                str(name) + "': type '" + 
                str(resolverType) + "' is not a PathValueResolver")

        if super(PathValueResolverRegistry, self).hasResolver(name) is False:
            resolver = resolverType() 
            resolver.registrationCallback(name, userData)
            self.registerResolver(resolver, name)
            self.__flatResolverStack[name] = resolver

    def getResolver(self, name):
        if name in __flatResolverStack:
            return self.__flatResolverStack[name]
        else:
            return super(PathValueResolverRegistry, self).getResolver(name)
             
    def unregisterFactory(self, name):
        if name in __flatResolverStack:
            del __flatResolverStack[name]
        super(PathValueResolverRegistry, self).unregisterFactory(name)

# \internal, store the registry singleton 
global s_pathValueResolverRegistrySingleton
s_pathValueResolverRegistrySingleton = None

def GetPathValueResolverRegistry():
    """ Creates the PathValueResolverRegistry singleton.
    """
    global s_pathValueResolverRegistrySingleton
    if s_pathValueResolverRegistrySingleton is None:
        s_pathValueResolverRegistrySingleton = PathValueResolverRegistry()
    return s_pathValueResolverRegistrySingleton
